### ISSUE TITLE
[mdns-mdnssd] potential access of freed `mServiceRef` from `Stop()`

### ISSUE THREAD

**Issue by abtink** [openthread/ot-br-posix#2007]
Notice this as I was checking the logs and code related to MDNSResponder `Mdsn::Publisher` implementaion.

- We use a common `mHostsRef` for all host registrations.
- The `Stop()` is as follows:
https://github.com/openthread/ot-br-posix/blob/8a280e3516641f707324bbc83059736d02f68955/src/mdns/mdns_mdnssd.cpp#L240-L265
- It swaps the `mHostRegistrations` with a empty local map which means the entries in the map are retained and freed at the end of this method (when the local `hostRegistrations` map goes out of scope) and the destructor `~DnssdHostRegistration()` is called for all entries
- But `Stop()` will already call `DNSServiceRefDeallocate(mHostsRef);` before that so the the ref is already freed.

------

Digging at history of code, noticed the model of using `swap()` was added in #1221 and before this the maps were `clear()`.
@superwhd can you please share background/reason for this? Thanks.

**Comment by superwhd** [comment]
In my impression I did this to ensure the `mServiceRegistrations` and `mHostRegistrations` are cleaned up only after `DNSServiceRefDeallocate`. 

Probably `DNSServiceRefDeallocate(mHostsRef)` triggers something that touches `mServiceRegistrations/mHostRegistrations` (it may crash if the containers are not empty) but I cannot find an evidence today.

**Comment by abtink** [comment]
> In my impression I did this to ensure the `mServiceRegistrations` and `mHostRegistrations`'s elements are destructed only after `DNSServiceRefDeallocate`.

But that is the problem, the destructor of `DnssdHostRegisteration` will use the already freed and unallocated `mHostsRef`.

The documentation mentioned this:
https://github.com/apple-oss-distributions/mDNSResponder/blob/806254210edec4cab01f794f6fd28658aa6ba59d/mDNSShared/dns_sd.h#L403-L416
```c++
   /* ...
     *
     * 4. Don't Double-Deallocate
     * Calling DNSServiceRefDeallocate(OpRef) for a particular operation's DNSServiceRef terminates
     * just that operation. Calling DNSServiceRefDeallocate(SharedRef) for the main shared DNSServiceRef
     * (the parent DNSServiceRef, originally created by DNSServiceCreateConnection(&SharedRef))
     * automatically terminates the shared connection *and* all operations that were still using it.
     * After doing this, DO NOT then attempt to deallocate any remaining subordinate DNSServiceRef's.
     * The memory used by those subordinate DNSServiceRef's has already been freed, so any attempt
     * to do a DNSServiceRefDeallocate (or any other operation) on them will result in accesses
     * to freed memory, leading to crashes or other equally undesirable results.
     * You can deallocate individual operations first and then deallocate the parent DNSServiceRef last,
     * but if you deallocate the parent DNSServiceRef first, then all of the subordinate DNSServiceRef's
     * are implicitly deallocated, and explicitly deallocating them a second time will lead to crashes.
     *

```

which mean when we call `DNSServiceRefDeallocate(mHostsRef)` it will invalidates all the child `aRecordRef` created under the shared `mHotsRef` but they are still being tracked by `HostRegisteration` and will be used in destructor call.

The better way is to clear the `HostRegisterations` map first to invoke all destructor and remove all `aRecordRef` and only then call `DNSServiceRefDeallocate(mHostsRef)`.

---------

> Probably `DNSServiceRefDeallocate(mHostsRef)` triggers something that touches `mServiceRegistrations/mHostRegistrations` (it may crash if the containers are not empty) but I cannot find an evidence today.

When we call `DNSServiceRefDeallocate(mHostsRef)` the MDNSResponder will immediatley free the underling ref and will ensure to never invoke any of the callback using this ref (or any of its associated child `aRecordRef`s).

There is one edge-case if there is already some earlier callbacks/events that is still pending where in `PublisherMDnsSd::Process(const MainloopContext &aMainloop)` method we added the corresponding `ServiceRef` to `readyServices` list and as we iterate through the list and call `DNSServiceProcessResult()`, it may trigger callback that cause some of the other existing `Refs`  in the `readyServices` get invalidated. I think we need to add code to protect and handle against this case in a general way (not just when we call `Stop()` but in whatever way a `Ref` may get invalidated). I will add such a mechanism in the code.

**Comment by abtink** [comment]
- Submitted #2013 which should help address this.

**Comment by abtink** [comment]
This is resolved by above PRs so we can close it.
